src/util/simplify_expr_int.cpp

```diff
@@ -1139,10 +1139,35 @@ bool simplify_exprt::simplify_extractbits(exprt &expr)
       svalue.substr(width-integer2size_t(start)-1,
                     integer2size_t(start-end+1));
 
-    expr = constant_exprt(extracted_value, expr.type());
+    constant_exprt result(extracted_value, expr.type());
+    expr.swap(result);
 
     return false;
   }
+  else if(expr.src().id() == ID_concatenation)
+  {
+    mp_integer offset = width;
+
+    forall_operands(it, expr.src())
+    {
+      mp_integer op_width = pointer_offset_bits(it->type(), ns);
+
+      if(op_width <= 0)
+        return true;
+
+      if(start + 1 == offset && end + op_width == offset)
+      {
+        exprt tmp = *it;
+        if(tmp.type() != expr.type())
+          tmp.make_typecast(expr.type());
+        expr.swap(tmp);
```

Comment:
The above is somewhat dangerous, as "typecast" does reinterpretation. However, I would likely consider a case where the typecast is needed to be malformed. I would probably not do the simplification in this scenario, to be conservative.